0000-richer-documents.rst

```diff
@@ -0,0 +1,460 @@
+Richer error documents
+======================
+
+.. proposal-number:: Leave blank. This will be filled in when the proposal is
+                     accepted.
+.. ticket-url:: Leave blank. This will eventually be filled with the
+                ticket URL which will track the progress of the
+                implementation of the feature.
+.. implemented:: Leave blank. This will be filled in with the first GHC version which
+                 implements the described feature.
+.. highlight:: haskell
+.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/0>`_.
+            **After creating the pull request, edit this file again, update the
+            number in the link, and delete this bold sentence.**
+.. sectnum::
+.. contents::
+
+Enrich GHC's document types (``Doc`` and ``SDoc``) with support for
+annotations of some arbitrary type ``a``, and use this new capability
+to emit error messages with bits of expressions, types and more
+embedded in the documents, as AST values, as opposed to their
+textual form which is not amenable to any form of inspection or
+post-processing.
+
+This proposal drew a lot of inspiration from the discussion at
+`ghc#8809 <https://gitlab.haskell.org/ghc/ghc/issues/8809>`_.
+
+Motivation
+----------
+
+GHC's whole pretty printing infrastructure is based around the following
+types:::
+
+    data Doc
+      = Empty                                            -- empty
+      | NilAbove Doc                                     -- text "" $$ x
+      | TextBeside !TextDetails {-# UNPACK #-} !Int Doc  -- text s <> x
+      | Nest {-# UNPACK #-} !Int Doc                     -- nest k x
+      | Union Doc Doc                                    -- ul `union` ur
+      | NoDoc                                            -- The empty set of documents
+      | Beside Doc Bool Doc                              -- True <=> space between
+      | Above Doc Bool Doc                               -- True <=> never overlap
+
+    data SDocContext = SDC { ... }
+    newtype SDoc = SDoc { runSDoc :: SDocContext -> Doc }
+
+The exact contents of ``SDocContext`` is not very important, one just needs
```

Comment:
Maybe just show `DynFlags` here?

Reply:
Well, the example at the end of the _Effects and interactions_ section uses the real definition. Did you get the impression that mentioning `SDocContext` adds a lot of "weight"?

Reply:
What I meant was to show that `SDocContext` includes `DynFlags`, among other things. I just thought it would help to illustrate what `SDocContext` is about.

Reply:
Perhaps I can just spell it out too, but not necessarily show the definition of `PprStyle` and what not. It's just one line to add.

Reply:
Sounds like a plan.